File: backend/02-profile/internal/observability/apm.go
// APM wiring with explicit runtime detection (docker / kubernetes / native).
// Mirrors dokandar-auth/app/observability/apm.py — the labels and
// service.node.name semantics are identical.
package observability

import (
	"fmt"
	"net"
	"net/url"
	"os"
	"strings"
	"time"

	"go.elastic.co/apm/v2"
)

// SetupAPM configures the default APM tracer with explicit runtime context.
// Must run BEFORE chi.Use(apmchiv5.Middleware(...)) so the tracer sees the
// labels on every transaction.
func SetupAPM(serviceName, serviceVersion, environment, serverURL, secretToken string) error {
	// Reconfigure the global tracer to talk to our APM server.
	if serverURL != "" {
		_ = os.Setenv("ELASTIC_APM_SERVER_URL", serverURL)
	}
	if secretToken != "" {
		_ = os.Setenv("ELASTIC_APM_SECRET_TOKEN", secretToken)
	}
	_ = os.Setenv("ELASTIC_APM_SERVICE_NAME", serviceName)
	_ = os.Setenv("ELASTIC_APM_SERVICE_VERSION", serviceVersion)
	_ = os.Setenv("ELASTIC_APM_ENVIRONMENT", environment)

	runtimeKind, nodeName, labels := DetectRuntime()
	_ = os.Setenv("ELASTIC_APM_SERVICE_NODE_NAME", nodeName)

	// Build the GLOBAL_LABELS env-string the agent reads (comma-separated k=v).
	kv := []string{}
	for k, v := range labels {
		kv = append(kv, fmt.Sprintf("%s=%s", k, v))
	}
	if len(kv) > 0 {
		_ = os.Setenv("ELASTIC_APM_GLOBAL_LABELS", strings.Join(kv, ","))
	}

	// Recreate the default tracer so the new env vars take effect.
	t, err := apm.NewTracerOptions(apm.TracerOptions{
		ServiceName:        serviceName,
		ServiceVersion:     serviceVersion,
		ServiceEnvironment: environment,
	})
	if err != nil {
		return fmt.Errorf("apm tracer: %w", err)
	}
	apm.SetDefaultTracer(t)

	_ = runtimeKind // logged by the caller
	return nil
}

// DetectRuntime returns (runtime_kind, service_node_name, labels). Same
// detection order as the Python apm.py: kubernetes → docker → native.
func DetectRuntime() (string, string, map[string]string) {
	labels := map[string]string{}

	if os.Getenv("KUBERNETES_SERVICE_HOST") != "" {
		pod := firstNonEmpty(os.Getenv("POD_NAME"), os.Getenv("HOSTNAME"), getHostname())
		ns := os.Getenv("POD_NAMESPACE")
		node := os.Getenv("NODE_NAME")
		labels["runtime"] = "kubernetes"
		if pod != "" {
			labels["k8s_pod_name"] = pod
		}
		if ns != "" {
			labels["k8s_namespace"] = ns
		}
		if node != "" {
			labels["k8s_node_name"] = node
		}
		return "kubernetes", pod, labels
	}

	if _, err := os.Stat("/.dockerenv"); err == nil {
		cid := readDockerContainerID()
		labels["runtime"] = "docker"
		labels["container_id"] = cid
		labels["container_runtime"] = "docker"
		return "docker", cid, labels
	}

	labels["runtime"] = "native"
	return "native", getHostname(), labels
}

func readDockerContainerID() string {
	if b, err := os.ReadFile("/etc/hostname"); err == nil {
		return strings.TrimSpace(string(b))
	}
	return getHostname()
}

func firstNonEmpty(vals ...string) string {
	for _, v := range vals {
		if v != "" {
			return v
		}
	}
	return ""
}

// APMServerReachable — TCP-probes the APM server URL for /health.
func APMServerReachable(serverURL string, timeout time.Duration) (bool, string) {
	u, err := url.Parse(serverURL)
	if err != nil {
		return false, err.Error()
	}
	port := u.Port()
	if port == "" {
		if u.Scheme == "https" {
			port = "443"
		} else {
			port = "80"
		}
	}
	addr := net.JoinHostPort(u.Hostname(), port)
	c, err := net.DialTimeout("tcp", addr, timeout)
	if err != nil {
		return false, err.Error()
	}
	_ = c.Close()
	return true, fmt.Sprintf("%s tcp-ok", addr)
}

File: backend/01-auth/app/observability/apm.py
"""Elastic APM integration.

Adds runtime-context detection (docker / kubernetes / native) so the agent
ships every transaction with enough metadata for Kibana APM to identify
how this instance is running:

  * `service.node.name` — set explicitly to the container ID (docker),
    the pod name (k8s), or the hostname (native). Kibana APM groups
    instances by this in the **Service overview → Instances** view.
  * `labels.runtime` — one of `docker` / `kubernetes` / `native`.
    Searchable, groupable, dashboardable.
  * `labels.container_id` (docker) or `labels.k8s_pod_name` /
    `labels.k8s_namespace` / `labels.k8s_node_name` (k8s).

Why labels and not the top-level `container.id` / `kubernetes.*` fields:
elastic-apm-python's auto-detector reads `/proc/self/cgroup`, which on
Ubuntu 26.04 cgroup v2 returns the unified `0::/` and the agent can't
extract a container ID from it. Labels are the public-API path; they
give the same UX in Kibana (filter, group, dashboard) — only the
"Container" sidebar icon is missing, which doesn't affect any query or
alert path.
"""
from __future__ import annotations
import logging
import os
import socket
from urllib.parse import urlparse

from elasticapm.contrib.starlette import ElasticAPM, make_apm_client

from app.config import settings

log = logging.getLogger("auth.apm")
_apm_client = None


def _detect_runtime() -> tuple[str, str, dict]:
    """Return (runtime_kind, service_node_name, labels_dict).

    Detection order: kubernetes -> docker -> native. K8s is checked
    first because a pod runs inside a docker (or other) container, but
    we want it tagged as `kubernetes`, not `docker`.
    """
    labels: dict[str, str] = {}

    if os.environ.get("KUBERNETES_SERVICE_HOST"):
        # K8s clusters set KUBERNETES_SERVICE_HOST in every pod. Pod
        # identity comes from env vars the manifest must inject via
        # downwardAPI (`fieldRef`): POD_NAME / POD_NAMESPACE / NODE_NAME.
        # Falls back to HOSTNAME (set to pod name by k8s) if the manifest
        # didn't add the downward fields.
        pod  = os.environ.get("POD_NAME") or os.environ.get("HOSTNAME") or socket.gethostname()
        ns   = os.environ.get("POD_NAMESPACE", "")
        node = os.environ.get("NODE_NAME", "")
        labels["runtime"] = "kubernetes"
        if pod:  labels["k8s_pod_name"]  = pod
        if ns:   labels["k8s_namespace"] = ns
        if node: labels["k8s_node_name"] = node
        return "kubernetes", pod, labels

    if os.path.exists("/.dockerenv"):
        # /.dockerenv is a sentinel file the docker engine creates in
        # every container. /etc/hostname inside a docker container is
        # the container ID (unless the user overrode it via --hostname).
        try:
            with open("/etc/hostname") as f:
                cid = f.read().strip()
        except Exception:
            cid = socket.gethostname()
        labels["runtime"] = "docker"
        labels["container_id"] = cid
        labels["container_runtime"] = "docker"
        return "docker", cid, labels

    # Bare native run on a VM or laptop.
    host = socket.gethostname()
    labels["runtime"] = "native"
    return "native", host, labels


def _labels_to_kv_string(labels: dict[str, str]) -> str:
    """elasticapm's GLOBAL_LABELS env-var format: comma-separated key=value.
    No commas or equals signs inside values; container IDs and pod names
    are safe (alphanumeric + dashes/dots)."""
    return ",".join(f"{k}={v}" for k, v in labels.items() if v)


def get_apm_client():
    global _apm_client
    if _apm_client is None:
        runtime_kind, node_name, labels = _detect_runtime()
        log.info(
            "apm runtime=%s node=%s labels=%s",
            runtime_kind, node_name, labels,
        )
        _apm_client = make_apm_client({
            "SERVICE_NAME":      settings.apm_service_name,
            "SERVER_URL":        settings.apm_server_url,
            "SECRET_TOKEN":      settings.apm_secret_token,
            "ENVIRONMENT":       settings.app_env,
            "SERVICE_VERSION":   settings.code_version,
            "SERVICE_NODE_NAME": node_name,
            "GLOBAL_LABELS":     _labels_to_kv_string(labels),
        })
    return _apm_client


def install(app):
    try:
        client = get_apm_client()
        app.add_middleware(ElasticAPM, client=client)
        log.info("apm installed (server=%s)", settings.apm_server_url)
    except Exception as e:
        log.warning("apm install failed: %s", e)


def health_check(timeout_s: float = 2.0) -> tuple[bool, str]:
    """TCP-level reachability check for the APM server URL.

    Returns (reachable, detail). Used by /ready and /health dep checks.
    Cheap — we don't want to spam the APM server with deep probes.
    """
    try:
        u = urlparse(settings.apm_server_url)
        host = u.hostname or "localhost"
        port = u.port or (443 if u.scheme == "https" else 80)
        with socket.create_connection((host, port), timeout=timeout_s):
            return True, f"{host}:{port} tcp-ok"
    except Exception as e:
        return False, str(e)[:80]
